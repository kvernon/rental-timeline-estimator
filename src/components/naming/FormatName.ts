/**
 * used for prefixing names
 * @param value
 * @param suffix
 */
export const formatName = (value: string, suffix: string): string => {
  const dot = '.';

  if (!value || value.indexOf(dot) === -1) return suffix + value;

  const strings = value.split(dot);

  strings[1] = strings[1] + dot + suffix;

  return strings.join(dot);
};
