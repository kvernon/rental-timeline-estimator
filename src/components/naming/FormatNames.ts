export const FormatNames = {
  ValidatorStackId: 'validatorStack',
  TitleDropDownValidatorId: 'titleDropDown',
  PropertyDropDownValidatorId: 'titleDropDown',
  RangeFieldValidatorId: 'rangeFieldValidator',
};
