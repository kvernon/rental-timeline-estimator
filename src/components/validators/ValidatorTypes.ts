export enum ValidatorTypes {
  Invalid = 0,
  Valid = 1,
  Optional = 2,
}
