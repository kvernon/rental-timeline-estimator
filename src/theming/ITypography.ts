import { IFontInformation } from './IFontInformation';
import { FontGroups } from './fontGroups';

export interface ITypography {
  parent: IFontInformation;

  get(key: FontGroups): IFontInformation | undefined;
}
