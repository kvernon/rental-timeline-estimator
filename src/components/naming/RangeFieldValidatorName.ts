import { formatName } from './FormatName';

export const RangeFieldValidatorId = 'rangeFieldValidator';
export const RangeFieldValidatorName = (id: string) => formatName(id, RangeFieldValidatorId);
