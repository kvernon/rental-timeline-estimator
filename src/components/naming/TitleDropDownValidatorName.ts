import { formatName } from './FormatName';
import { FormatNames } from './FormatNames';

export const TitleDropDownValidatorName = (id: string) => formatName(id, FormatNames.TitleDropDownValidatorId);
