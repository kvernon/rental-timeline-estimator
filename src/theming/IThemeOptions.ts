import { IPalette } from './IPalette';

import { ITypography } from './ITypography';

export interface IThemeOptions {
  palette: IPalette;
  typography: ITypography;
}
