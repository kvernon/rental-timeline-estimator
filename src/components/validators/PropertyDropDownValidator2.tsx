import React, { ReactNode, useEffect, useState } from 'react';
import { Controller, useFormContext } from 'react-hook-form';
import ReactSelect, { GroupBase, SingleValue } from 'react-select';
import { useTheme } from '@emotion/react';
import { IThemeOptions } from '../../theming/IThemeOptions';
import { ValidatorTypes } from './ValidatorTypes';
import { FontGroups } from '../../theming/fontGroups';
import styled from '@emotion/styled';
import { formatName } from '../naming/FormatName';
import { FormatNames } from '../naming/FormatNames';

export interface IPropertyDropDownParams {
  id?: string;
  onChange?: (value: IPropertyDropDownOption) => void;
}

export interface IPropertyDropDownOption {
  label: string;
  value: number;
  image: string;
}

const Img = styled.img`
  padding: 0;
`;

export function PropertyDropDownValidator2<
  Option extends IPropertyDropDownOption,
  IsMulti extends boolean = false,
  Group extends GroupBase<Option> = GroupBase<Option>,
>(props: IPropertyDropDownParams) {
  const formatOptionLabel = (data: IPropertyDropDownOption): ReactNode => {
    return <Img src={data.image} alt={data.label} title={data.label} />;
  };

  const { control, setValue } = useFormContext();
  const coreTheme = useTheme() as IThemeOptions;

  const selectUuid = props.id || window.crypto.randomUUID();
  const inputValidationValue = `${formatName(selectUuid, FormatNames.PropertyDropDownValidatorId)}.value`;
  const inputValidationResult = `${formatName(selectUuid, FormatNames.PropertyDropDownValidatorId)}.validationResult`;

  const optionsMap = ['apartment', 'house'].map((title: string, idx: number) => {
    return { value: idx, label: title, image: `/images/${title}.jpg` };
  }) as (Option | Group)[];

  const [selectedIndex, setSelectedIndex] = useState<number>(0);

  useEffect(() => {
    setValue(inputValidationValue, selectedIndex, { shouldDirty: true, shouldValidate: true, shouldTouch: true });
    setValue(inputValidationResult, ValidatorTypes.Valid);
  }, []);

  const handleChange = (option: SingleValue<IPropertyDropDownOption | unknown>): void => {
    const value = (option as IPropertyDropDownOption).value;
    if (value !== selectedIndex) {
      setValue(inputValidationValue, option);
      setSelectedIndex(value);
    }
  };

  const Select = styled(ReactSelect<Option, IsMulti, Group>)<{
    themeOptions: IThemeOptions;
  }>`
    appearance: none;
    white-space: pre-wrap;
    font-size: ${(props) => props.themeOptions.typography.get(FontGroups.inputLabel)?.size};
    font-family: ${(props) => props.themeOptions.typography.get(FontGroups.inputLabel)?.font};
    font-weight: ${(props) => props.themeOptions.typography.get(FontGroups.inputLabel)?.weight};
    width: 100%;
    padding-left: 10px;
    color: ${(props) => props.themeOptions.typography.get(FontGroups.input)?.color};
    overflow: visible;
  `;

  return (
    <Controller
      name={inputValidationValue}
      control={control}
      render={({ field }) => {
        return (
          <Select
            {...field}
            defaultValue={optionsMap[selectedIndex] as Option}
            formatOptionLabel={formatOptionLabel}
            themeOptions={coreTheme}
            options={optionsMap}
            onChange={handleChange}
            styles={{
              singleValue: (base) => ({
                ...base,
                color: `${coreTheme.typography.get(FontGroups.input)?.color}`,
              }),
              menu: (baseStyles) => ({
                ...baseStyles,
                zIndex: baseStyles.isSelected ? 9999 : baseStyles.zIndex,
              }),
              control: (baseStyles) => {
                return {
                  ...baseStyles,
                  overflow: 'visible',
                  transition: 'background-color 0.4s ease-out',
                  backgroundColor: `${coreTheme.palette.validation[ValidatorTypes[ValidatorTypes.Valid]].background}41`,
                  height: '59px',
                  borderColor: `${coreTheme.palette.inputBackground}`,
                  border: `1px solid ${coreTheme.palette.panelBackground}`,
                  borderRadius: '0.3rem',
                  color: `${coreTheme.typography.get(FontGroups.input)?.color}`,
                  ':hover': {
                    backgroundColor: `${coreTheme.palette.validation[ValidatorTypes[ValidatorTypes.Valid]].background}81`,
                    borderColor: `${coreTheme.palette.inputBackgroundFocus}`,
                    color: `${coreTheme.palette.inputBackgroundFocus}`,
                  },
                };
              },
            }}
          />
        );
      }}
    />
  );
}
