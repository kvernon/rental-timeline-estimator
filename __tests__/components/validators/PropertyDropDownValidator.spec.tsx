import { configure, Matcher, render, screen } from '@testing-library/react';
import React from 'react';
import { formatName } from '../../../src/components/naming/FormatName';
import { FormatNames } from '../../../src/components/naming/FormatNames';
import { IPropertyDropDownParams, PropertyDropDownValidator2 } from '../../../src/components/validators/PropertyDropDownValidator2';
import { FormProvider, useForm } from 'react-hook-form';
import { ValidatorTypes } from '../../../src/components/validators/ValidatorTypes';
import { Theme, useTheme } from '@emotion/react';
import { IThemeOptions } from '../../../src/theming/IThemeOptions';
import { ITypography } from '../../../src/theming/ITypography';
import selectEvent from 'react-select-event';
import '@testing-library/jest-dom';

const Setup = (props: IPropertyDropDownParams) => {
  const methods = useForm({
    mode: 'onBlur',
    defaultValues: {
      [`${formatName(props.id as string, FormatNames.PropertyDropDownValidatorId)}`]: {
        value: {
          value: 0,
          label: 'apartment',
        },
        validationResult: ValidatorTypes.Valid,
      },
    },
  });

  return (
    <FormProvider {...methods}>
      <PropertyDropDownValidator2 {...props} />
    </FormProvider>
  );
};

jest.mock('@emotion/react', () => {
  const requireActual = jest.requireActual('@emotion/react');
  const useTheme: jest.MockedFn<() => Theme> = jest.fn();
  return {
    ...requireActual,
    useTheme,
  };
});

describe('PropertyDropDownValidator unit test', () => {
  let typographyMock: jest.Mocked<ITypography>;
  let params: IPropertyDropDownParams;

  const validationColorOptionalRight = '#0000FF';
  const validationColorValidMiddle = '#00FF00';
  const validationColorInvalidLeft = '#FF0000';

  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    configure({ testIdAttribute: 'name' });
    const useThemeMock = jest.mocked(useTheme);
    typographyMock = {
      parent: {
        allPopulated: jest.fn(),
        font: 'p',
        color: 'p',
        size: 'p',
      },
      get: jest.fn().mockReturnValue({
        font: 'child',
        color: 'child',
        size: 'child',
      }),
    };

    useThemeMock.mockReturnValue({
      palette: {
        pageBackground: 'pageBackground',
        panelBackground: 'panelBackground',
        panelShadow: 'panelShadow',

        inputBackgroundBadFocus: 'inputBackgroundBadFocus',
        inputBackgroundBad: 'inputBackgroundBad',
        inputBackground: 'inputBackground',
        inputBackgroundFocus: 'inputBackgroundFocus',

        validation: {
          Invalid: {
            validationColor: validationColorInvalidLeft,
            background: validationColorInvalidLeft,
            backgroundFocus: validationColorInvalidLeft,
          },
          Valid: {
            validationColor: validationColorValidMiddle,
            background: validationColorValidMiddle,
            backgroundFocus: validationColorValidMiddle,
          },
          Optional: {
            validationColor: validationColorOptionalRight,
            background: validationColorOptionalRight,
            backgroundFocus: validationColorOptionalRight,
          },
        },
      },
      typography: typographyMock,
    } as jest.Mocked<IThemeOptions>);

    params = {
      id: 'Tested',
    };
  });

  describe('and defaults', () => {
    test('should exist', () => {
      render(<Setup {...params} />);

      const entity = screen.queryByTestId<HTMLInputElement>(`${formatName(params.id as string, FormatNames.PropertyDropDownValidatorId)}.value`);

      expect(entity).toMatchSnapshot();
    });

    test('should contain titles', () => {
      render(<Setup {...params} />);

      const entity = screen.getByTestId<HTMLElement>(`${formatName(params.id as string, FormatNames.PropertyDropDownValidatorId)}.value`);

      selectEvent.openMenu(entity);

      const allByText = screen.getAllByRole<HTMLImageElement>('img');

      expect(
        allByText.map((x) => ({
          src: x.src,
          title: x.title,
          alt: x.alt,
        })),
      ).toEqual(
        ['apartment', 'house'].map((exp) => ({
          src: `http://localhost/images/${exp}.jpg`,
          title: exp,
          alt: exp,
        })),
      );
    });

    /*    test('call change', async () => {
      render(<Setup {...params} />);


      const choices = ['apartment', 'house'].map((exp) => ({
        src: `http://localhost/images/${exp}.jpg`,
        title: exp,
        alt: exp,
      }));

      const entity = screen.getByTestId<HTMLInputElement>(`${formatName(params.id as string, FormatNames.PropertyDropDownValidatorId)}.value`);

      selectEvent.openMenu(entity);

      const element = screen.getByAltText<HTMLImageElement>('house');

      const expectedValue = choices[1];

      console.log(element);
      await selectEvent.select(entity);

      expect(screen.getByText('selectedOption')).toBeInTheDocument();
      expect(screen.getByTestId<HTMLInputElement>(params.id as string).value).toEqual(expectedValue.toString());
    });*/
  });
});
