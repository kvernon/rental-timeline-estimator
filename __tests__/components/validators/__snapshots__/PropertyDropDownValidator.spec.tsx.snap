// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyDropDownValidator unit test and defaults should exist 1`] = `
<input
  name="titleDropDownTested.value"
  type="hidden"
  value=""
/>
`;
